# problem 1
# (a)
library(ALL)
data(ALL)
ALL.fac <- as.numeric(ALL$BT)
ALL.fac[ALL.fac <= 5] = 1
ALL.fac[ALL.fac > 5] = 2

# (b)
par(mfrow=c(1,3))
for (i in 1:3) {
  hist(exprs(ALL)[i,], main="gene expression", nclass=15, 
       freq=FALSE, xlab=rownames(exprs(ALL))[i])
}

# (c)
firstFive <- t(exprs(ALL)[1:5,])
pairs(firstFive, col=ALL.fac)

# (d)
rowname <- rownames(exprs(ALL))
index39317 <- grep("^39317_at$", rowname)
index32649 <- grep("^32649_at$", rowname)
index481 <- grep("^481_at$", rowname)
data <- t(exprs(ALL)[c(index39317, index32649, index481),])

require(scatterplot3d)
scatterplot3d(data, color=ALL.fac)

# (e)
cl.2mean <- kmeans(data, centers=2, nstart = 10)
table(ALL.fac, cl.2mean$cluster)

cl.3mean <- kmeans(data, centers=3, nstart = 10)
table(ALL.fac, cl.3mean$cluster)

# (f)
PCA <- prcomp(exprs(ALL), scale=TRUE)
summary(PCA)

# (g)
print(c(t(PCA$rotation[,1])), digits=3)
biplot(PCA, xlim=c(-0.05,0.03), ylim=c(-0.05,0.05), cex=0.5)

# (h)
o <- order(PCA$x[,2])

numOfGenes <- dim(exprs(ALL))[1]
rowname[ o[ (numOfGenes-2) : numOfGenes]]

rowname[o[1:3]]

# (i)
library(hgu95av2.db)
biggest <- rowname[o[numOfGenes]]
smallest <- rowname[o[1]]

# gene name and chromosome for the gene with biggest PC2 value
get(biggest, env = hgu95av2GENENAME)
get(biggest, env = hgu95av2CHR)

# gene name and chromosome for the gene with smallest PC2 value
get(smallest, env = hgu95av2GENENAME)
get(smallest, env = hgu95av2CHR)



# problem 2
rm(list=ls())
# (a)
data <- data.frame(iris[,1:4])
scaled.data <- scale(data)

colMeans(scaled.data)
apply(scaled.data, 2, sd)

# (b)
cor(data)
cor(scaled.data)

# (c)
dist(t(scaled.data), method="eucl")
d <- c(dist(t(scaled.data), method="eucl")^2)
corr <- cor(scaled.data)[lower.tri(diag(4))]
d/(1-corr)

# (d)
unscaled.PCA <- prcomp(data, scale=FALSE)
scaled.PCA <- prcomp(scaled.data, scale=FALSE)

summary(scaled.PCA)
summary(unscaled.PCA)

# (e)
#scaled: 
#PC1: 72.96% 
#PC2: 22.85%
#PC1+PC2: 95.81%

#unscaled: 
#PC1: 92.46% 
#PC2: 5.307%
#PC1+PC2: 97.769%

# (f)
p <- ncol(scaled.data)
n <- nrow(scaled.data) 
nboot<-1000 
sdevs <- array(dim=c(nboot,p)) 
for (i in 1:nboot) { 
  dat.star <- scaled.data[sample(1:n,replace=TRUE),] 
  sdevs[i,] <- (prcomp(dat.star)$sdev)^2  
} 
as.numeric(quantile(sdevs[,2]/apply(sdevs, 1, sum), c(0.05,0.95)))